File: go/agent-runner/adapters/codex.go
package adapters

import (
	"os/exec"
	"regexp"
	"strconv"
	"strings"

	runner "github.com/qwegle/setra/agent-runner"
)

// Codex implements the Adapter interface for OpenAI's Codex CLI.
type Codex struct{}

func (c *Codex) Name() string        { return "codex" }
func (c *Codex) DisplayName() string  { return "OpenAI Codex" }
func (c *Codex) DefaultModel() string { return "o4-mini" }

func (c *Codex) SupportedModels() []string {
	return []string{"o4-mini", "o3", "gpt-4.1"}
}

func (c *Codex) IsAvailable() bool {
	_, err := exec.LookPath("codex")
	return err == nil
}

func (c *Codex) BuildCommand(opts runner.BuildCommandInput) runner.SpawnOptions {
	args := []string{"-q", opts.Task}
	if opts.Model != "" {
		args = append(args, "--model", opts.Model)
	}
	return runner.SpawnOptions{
		Command: "codex",
		Args:    args,
		WorkDir: opts.WorkDir,
	}
}

var (
	codexPromptRegex = regexp.MustCompile(`prompt tokens:\s*([0-9,]+)`)
	codexCompRegex   = regexp.MustCompile(`completion tokens:\s*([0-9,]+)`)
	codexCostRegex   = regexp.MustCompile(`cost:\s*\$([0-9.]+)`)
)

func (c *Codex) ParseTokenUsage(output string) (int64, int64) {
	var inp, out int64
	if m := codexPromptRegex.FindStringSubmatch(output); len(m) >= 2 {
		inp = parseTokenCount(m[1])
	}
	if m := codexCompRegex.FindStringSubmatch(output); len(m) >= 2 {
		out = parseTokenCount(m[1])
	}
	return inp, out
}

func (c *Codex) ParseCostUSD(output string) float64 {
	m := codexCostRegex.FindStringSubmatch(output)
	if len(m) < 2 {
		return 0
	}
	cost, _ := strconv.ParseFloat(m[1], 64)
	return cost
}

func (c *Codex) DetectCompletion(output string) bool {
	return strings.Contains(output, "cost:")
}

func (c *Codex) DetectRateLimit(output string) bool {
	return strings.Contains(output, "rate_limit") || strings.Contains(output, "429")
}

File: packages/agent-runner/src/adapters/codex.ts
/**
 * CodexAdapter — wraps the OpenAI Codex CLI (`codex exec` subcommand).
 *
 * Verified flags for codex-cli v0.128+:
 *   exec                                      Non-interactive subcommand
 *   -m, --model <id>                          Any OpenAI model (gpt-4o, gpt-5.5, o3, …)
 *   --dangerously-bypass-approvals-and-sandbox Skip all confirmations (full-auto equivalent)
 *   -C <dir>                                  Working directory
 *   --skip-git-repo-check                     Allow running outside a git repo
 *   --color never                             No ANSI color in output
 *
 * Note: --approval-mode, --quiet, and --mcp-config do NOT exist in v0.128.
 * MCP servers must be pre-configured in ~/.codex/config.toml.
 */

import { execFileSync } from "child_process";
import type { AgentAdapter } from "../adapter.js";
import type { Plot, Run, SpawnOptions, TokenUsage } from "../types.js";

export const CODEX_RATE_LIMIT_PATTERNS: readonly RegExp[] = [
	/rate[\s_-]?limit/i,
	/\b429\b/,
	/quota\s+exceeded/i,
	/too\s+many\s+requests/i,
	/insufficient_quota/i,
	/billing/i,
] as const;

const CODEX_COMPLETION_PATTERNS: readonly RegExp[] = [
	/Task\s+complete\./i,
	/cost:\s+\$[\d.]+/i,
] as const;

// eslint-disable-next-line no-control-regex
const STRIP_ANSI_RE = /\x1b\[[0-9;]*[mGKHFABCDJsu]/g;
const CODEX_COST_RE = /cost:\s+\$?([\d.]+)/i;
const CODEX_PROMPT_TOKENS_RE = /prompt tokens:\s+([\d,]+)/i;
const CODEX_COMPLETION_TOKENS_RE = /completion tokens:\s+([\d,]+)/i;
const CODEX_CACHED_TOKENS_RE = /cached tokens:\s+([\d,]+)/i;

export class CodexAdapter implements AgentAdapter {
	readonly name = "codex" as const;
	readonly displayName = "Codex CLI (OpenAI)";
	// Open-ended: codex supports any OpenAI model; we pass the ID through as-is.
	readonly supportsModels = [] as const;
	readonly defaultModel = "gpt-4o";

	async isAvailable(): Promise<boolean> {
		try {
			execFileSync("which", ["codex"], { stdio: "pipe" });
			return true;
		} catch {
			return false;
		}
	}

	buildCommand(plot: Plot, run: Run, _mcpConfigPath: string): SpawnOptions {
		const model = run.model?.trim() || this.defaultModel;

		// codex exec -m <model> --dangerously-bypass-approvals-and-sandbox
		//   --skip-git-repo-check --color never -C <cwd> <task>
		const args: string[] = [
			"exec",
			"-m",
			model,
			"--dangerously-bypass-approvals-and-sandbox",
			"--skip-git-repo-check",
			"--color",
			"never",
		];

		if (plot.worktreePath) {
			args.push("-C", plot.worktreePath);
		}

		if (run.systemPromptAppend) {
			args.push(...this.buildSystemPromptArgs(run.systemPromptAppend));
		}

		args.push(run.task);

		return {
			cmd: "codex",
			args,
			env: {
				SETRA_PLOT_ID: plot.id,
				SETRA_RUN_ID: run.id,
				SETRA_AGENT: this.name,
				SETRA_MODEL: model,
			},
			// cwd still set so shell inherits it; -C is the codex-level override
			cwd: plot.worktreePath,
		};
	}

	buildSystemPromptArgs(systemPrompt: string): string[] {
		// codex exec supports --instructions for system-level context
		return ["--instructions", systemPrompt];
	}

	// MCP is configured globally in ~/.codex/config.toml — no CLI flag in v0.128
	buildMcpArgs(_mcpConfigPath: string): string[] {
		return [];
	}

	parseTokenUsage(output: string): TokenUsage | null {
		const clean = output.replace(STRIP_ANSI_RE, "");
		const parseNum = (s: string | undefined) =>
			Number.parseInt((s ?? "0").replace(/,/g, ""), 10);

		const promptMatch = CODEX_PROMPT_TOKENS_RE.exec(clean);
		const completionMatch = CODEX_COMPLETION_TOKENS_RE.exec(clean);
		if (!promptMatch || !completionMatch) return null;

		return {
			promptTokens: parseNum(promptMatch[1]),
			completionTokens: parseNum(completionMatch[1]),
			cacheReadTokens: parseNum(CODEX_CACHED_TOKENS_RE.exec(clean)?.[1]),
			cacheWriteTokens: 0,
		};
	}

	parseCostUSD(output: string): number | null {
		const clean = output.replace(STRIP_ANSI_RE, "");
		const match = CODEX_COST_RE.exec(clean);
		if (!match) return null;
		const cost = Number.parseFloat(match[1] ?? "NaN");
		return isNaN(cost) ? null : cost;
	}

	detectRateLimit(output: string): boolean {
		return CODEX_RATE_LIMIT_PATTERNS.some((p) => p.test(output));
	}

	detectCompletion(output: string): boolean {
		return CODEX_COMPLETION_PATTERNS.some((p) => p.test(output));
	}
}

export const codexAdapter = new CodexAdapter();
